Source/PZ_C_2/Weapon/BaseWeapon.cpp

```diff
@@ -0,0 +1,90 @@
+// Fill out your copyright notice in the Description page of Project Settings.
+
+
+#include "BaseWeapon.h"
+
+#include "CollisionQueryParams.h"
+#include "TimerManager.h"
+#include "Engine/SkeletalMeshSocket.h"
+
+UBaseWeapon::UBaseWeapon()
+{
+	Range = 10000;
+	MuzzleSocketName = FName("Muzzle");
+}
+
+
+void UBaseWeapon::Fire()
+{
+	if( !this->CanFire() ){
+		return;
+	}
+	
+	USkeletalMeshSocket* Socket = SkeletalMesh->FindSocket(this->MuzzleSocketName);
+	if(Socket == nullptr){
+		return;
+	}
+	
+	FVector SocketLocation = Socket->GetSocketLocation(this);
+	GEngine->AddOnScreenDebugMessage(-1, 5.f, FColor::Green, SocketLocation.ToString());
+
+	FVector TraceEndLocation = SocketLocation + (GetOwner()->GetActorForwardVector() * Range);
+	WeaponTrace(SocketLocation, TraceEndLocation);
+	UseAmmo();
+}
+
+bool UBaseWeapon::bCanReload() const
+{
+	return !IsReloading && AmmoInClip < MaxAmmoInClip && AmmoTotal > 0; 
+}
+
+bool UBaseWeapon::CanFire() const
+{
+	return AmmoInClip > 0 && !IsReloading;
+}
+
+void UBaseWeapon::Reload()
+{
+	if( !bCanReload() || IsReloading ){
+		return;
+	}
+
+	IsReloading = true;
+	GEngine->AddOnScreenDebugMessage(-1, 5.f, FColor::Green, TEXT("Reloading"));
+
+	FTimerHandle UnusedHandle;
+	FTimerDelegate TimerCallback;
+	TimerCallback.BindLambda([&] 
```

Comment:
многовато функционала для лямбды. из стандартов:
- The best lambdas should be no more than a couple of statements in length
- Explicit captures should be used rather than automatic capture ([&] and [=])

лучше использовать отдельный метод.